tests/sepp/Execution.cpp

```diff
@@ -0,0 +1,144 @@
+///
+/// @author Manuel Pöter
+////////////////////////////////////////////////////////////////////////////////
+
+#include "Execution.h"
+
+#include <iostream>
+#include <filesystem>
+#include <memory>
+
+#include "Workload.h"
+
+namespace {
+std::size_t getFolderSize(std::string_view path) {
+  return std::accumulate(std::filesystem::recursive_directory_iterator(path),
+                         std::filesystem::recursive_directory_iterator(), 0ull,
+                         [](auto size, auto& path) {
+                           return std::filesystem::is_directory(path)
+                                      ? size
+                                      : size + std::filesystem::file_size(path);
+                         });
+}
+}  // namespace
+
+namespace arangodb::sepp {
+
+Execution::Execution(Options const& options, std::shared_ptr<Workload> workload)
+    : _state(ExecutionState::starting),
+      _options(options),
+      _workload(std::move(workload)) {}
+
+Execution::~Execution() {
+  _state.store(ExecutionState::stopped);
+  for (auto& thread : _threads) {
+    if (thread->_thread.joinable()) {
+      thread->_thread.join();
+    }
+  }
+}
+
+void Execution::createThreads(Server& server) {
+  _threads = _workload->createThreads(*this, server);
+}
+
+ExecutionState Execution::state(std::memory_order order) const {
+  return _state.load(order);
+}
+
+void Execution::signalFinishedThread() noexcept { _activeThreads.fetch_sub(1); }
+
+Report Execution::run() {
+  _activeThreads.store(_threads.size());
+  _state.store(ExecutionState::preparing);
+
+  waitUntilAllThreadsAre(ThreadState::running);
+
+  _state.store(ExecutionState::initializing);
+
+  waitUntilAllThreadsAre(ThreadState::ready);
+
+  _state.store(ExecutionState::running);
+
+  auto start = std::chrono::high_resolution_clock::now();
+
+  std::uint32_t sleeptimePerRound = 100;
+  auto rounds = [&]() -> std::uint32_t {
+    auto stoppingCriterion = _workload->stoppingCriterion();
+    if (std::holds_alternative<StoppingCriterion::Runtime>(stoppingCriterion)) {
+      return std::get<StoppingCriterion::Runtime>(stoppingCriterion).ms /
+             sleeptimePerRound;
+    } else {
+      // for stopping criteria based on number of operations we use an artifical
+      // time limit of one hour
+      return 1000 * 60 * 60 / 100;
+    }
+  }();
+
+  for (std::uint32_t r = 0;
+       r < rounds && _activeThreads.load(std::memory_order_relaxed) > 0; ++r) {
+    std::this_thread::sleep_for(std::chrono::milliseconds(sleeptimePerRound));
+    // TODO - record samples (memory usage, rocksdb stats, ...)
+  }
+
+  _state.store(ExecutionState::stopped);
+
+  waitUntilAllThreadsAre(ThreadState::finished);
+
+  std::chrono::duration<double, std::milli> runtime =
+      std::chrono::high_resolution_clock::now() - start;
+  return buildReport(runtime.count());
+}
+
+Report Execution::buildReport(double runtime) {
+  for (auto& thread : _threads) {
+    thread->_thread.join();
```

Comment:
The join method used here is different to what's used in the destructor:
https://github.com/arangodb/arangodb/pull/16250/files?w=1#diff-81785ac87c72ccdc01f12636ec26ad54a3eeab43c8da68e233250ed9b39a4492R53-R57